bpo-46480: add typing.assert_type (#30843)
python/cpython: Doc/library/typing.rst

   runtime we intentionally don't check anything (we want this
   to be as fast as possible).

> .. function:: assert_type(val, typ, /)

Reviewer: My 2c is that this should go somewhere at the bottom of the page. Why? Because it is only interesting for advanced users (type-checker authors / stubs maintainers / libs authors).

It might confuse people a lot if it is "advertised" in the same section as regular typing features.

What do others think?

Reviewer: Maybe something like "Types maintenance helpers"? Or similar?

Author: We already have the similarly obscure `@type_check_only` in this section, along with `@no_type_check_decorator` which I'm pretty sure nobody uses. I could get on board with moving it near the bottom of the section, though.

   Assert (to the type checker) that the value is of the given type.

   When the type checker encounters a call to ``assert_type()``, it
   emits an error if the value is not of the specified type::

       def greet(name: str) -> None:
           assert_type(name, str)  # ok
           assert_type(name, int)  # type checker error

>    At runtime this returns the first argument unchanged.

Reviewer 2: Maybe clarify that there is no other _runtime_ effect (such as checking the value's type):

* At runtime this returns the first argument unchanged **and otherwise does nothing.**

Author: Good point, done

   runtime we intentionally don't check anything (we want this
   to be as fast as possible).

.. function:: assert_type(val, typ, /)

>    Assert (to the type checker) that the value is of the given type.

Reviewer 3: ```suggestion
   Assert (to the type checker) that *value* has a revealed type of *typ*.
```

The type checker has no knowledge of the _actual type_ of `value`. It only has knowledge of its _inferred type_ of `value`, which is what this function tests.

Author: I don't think "revealed type" makes sense if we're not calling reveal_type():

```suggestion
   Assert (to the type checker) that *value* has an inferred type of *type*.
```

   When the type checker encounters a call to ``assert_type()``, it
   emits an error if the value is not of the specified type::

       def greet(name: str) -> None:

>            assert_type(name, str)  # ok

Reviewer 3: ```suggestion
           assert_type(name, str)  # OK, revealed type of `name` is `str`
```

Author: ```suggestion
           assert_type(name, str)  # OK, inferred type of `name` is `str`
```

           assert_type(name, int)  # type checker error

>    At runtime this returns the first argument unchanged and otherwise
>    does nothing.

Reviewer 3: ```suggestion
   At runtime this returns the first argument unchanged with no side effects.
```

> 

Reviewer 3: ```suggestion

   This function is useful for ensuring the type checker's understanding of a
   script is in line with the developer's intentions::

       def complex_function(arg: object):
           # Do some complex type-narrowing logic,
           # after which we hope the inferred type will be `int`
           ...
           # Test whether the type checker correctly understands our function
           assert_type(arg, int)

```

(Or some other example along these lines)

   runtime we intentionally don't check anything (we want this
   to be as fast as possible).

.. function:: assert_type(val, typ, /)

>    Assert (to the type checker) that *value* has an inferred type of *type*.

Reviewer 3: ```suggestion
   Assert (to the type checker) that *value* has an inferred type of *typ*.
```
The parameter name is `typ`, not `type` :)

Author: Good catch! Also, the other param is named `val`, not `value`. I chose those names because `cast()` uses them too.
```suggestion
   Assert (to the type checker) that *val* has an inferred type of *typ*.
```
